pkg/runtime.v2/framework/plugins/plainml/plainml.go

```diff
@@ -40,16 +41,34 @@ func (p *PlainML) Name() string {
 	return Name
 }
 
-func (p *PlainML) EnforceMLPolicy(info *runtime.Info) error {
+func (p *PlainML) EnforceMLPolicy(info *runtime.Info, trainJob *kubeflowv2.TrainJob) error {
 	if info == nil || info.MLPolicy == nil || info.MLPolicy.Torch != nil || info.MLPolicy.MPI != nil {
 		return nil
 	}
-	numNodes := ptr.Deref(info.MLPolicy.NumNodes, 1)
+
+	// TrainJob contains the actual information for the number of nodes.
+	numNodes := info.MLPolicy.NumNodes
+	if trainJob.Spec.Trainer != nil && trainJob.Spec.Trainer.NumNodes != nil {
+		numNodes = trainJob.Spec.Trainer.NumNodes
+	}
+	info.Trainer.NumNodes = numNodes
+
+	// Add envs from the TrainJob.
+	if trainJob.Spec.Trainer != nil {
+		info.Trainer.Env = append(info.Trainer.Env, trainJob.Spec.Trainer.Env...)
```

Comment:
What if there are conflicts between info.trainer.env and trainJob.spec.trainer.env?

Reply:
For the PlainML `info.trainer.env` is always equal to `trainjob.spec.trainer.env`.
For the Torch `info.trainer.env` takes precedence over `trainjob.spec.trainer.env`.
We should validate that TrainJob doesn't include `PET_` envs: https://github.com/kubeflow/training-operator/blob/f00c1d37ef03d89214ddd9a7189f98f81fce2276/pkg/runtime.v2/framework/plugins/torch/torch.go#L68.
cc @akshaychitneni

Reply:
My concern was the situations where another plugin (upstream or third-party) insert the arbitrary env.
So, could you verify here the conflicts or just lead TODO comments?

Reply:
I see, I think in that case the plugin developer needs to make sure that envs are not conflicting.
I can make this as a TODO statement in the torch plugin to discuss it in the future.